Add the Climb along line algorithm (#3865)
qgis/QGIS-Documentation: source/docs/user_manual/processing_algs/qgis/vectoranalysis.rst

  HTML file with calculated statistics.

.. _qgisclimbalongline:

Climb along line
----------------

This algorithm calculates the total climb and descent along line geometries.
The input layer must have Z values present. If Z values are not available,

> the  "Drape" (set Z value from raster) algorithm may be used to add Z values

Author: ```suggestion
the  :ref:`qgissetzfromraster` algorithm may be used to add Z values
```

> the  "Drape" (set Z value from raster) algorithm may be used to add Z values

Author: Added

from a DEM layer.

The output layer is a copy of the input layer with additional fields that
contain the total climb, total descent, the minimum elevation and the
maximum elevation for each line geometry.

> If the input layer contains fields with the same names as these added fields,

Author: Overwriting was the original behaviour of the algorithm. I think the current behaviour was suggested by @nyalldawson. As long as the calculated values are stored in the specified attributes there should at least be no problems.

they will be renamed (field names will be altered to "name_2", "name_3", etc,
finding the first non-duplicate name).

> ``Default menu``: :menuselection:`Vector --> Analysis Tools`

Author: ```suggestion
```

> 

Author: ```suggestion
```

> ``Default menu``: :menuselection:`Vector --> Analysis Tools`

Author: I just copied the "Default menu" line from another description (Count points in polygon).

> ``Default menu``: :menuselection:`Vector --> Analysis Tools`

Reviewer: Actually this line means that on any install, if I go to that menu, by default I should find the tool (this is used as shortcut for most of old ftools/gdal tools and mainly for most commonly used algorithms). If it is not the case, it should be removed.

> ``Default menu``: :menuselection:`Vector --> Analysis Tools`

Author: I have now removed it (I thought it had something to do with the Processing toolbox hierarchy).

> ``Default menu``: :menuselection:`Vector --> Analysis Tools`

Reviewer: The Processing toolbox hierarchy is already followed, in the file name and folder.

Parameters
..........

``Line layer`` [vector: line]

>   Line layer to calculate the climb for.

Author: ```suggestion
  Line layer to calculate the climb for. Must have Z values.
```

>   Line layer to calculate the climb for.

Author: Done.

Outputs
.......
``Climb layer`` [vector: line]
  Line layer containing new attributes with the results from climb calculations.

> ``Total climb``

Author: These are algorithm outputs.  The numbers stored in the fields of the output layer is for each line.  These numbers are the over all sums of the numbers in the output layer.

> ``Total climb``

Author: ```suggestion
``Total climb`` [number]
```

>   The sum of the climb for all the line geometries in the layer.

Reviewer: I did have a look with latest build and I have questions/suggestions with following options. In the dialog, only `Climb layer` is the "entry". The other values are shown in the log panel. I wouldn't have found them if I were not looking for them. What about another output entry in the "parameters" tab that users can check if they want to output these global values? It could be a html file or whatever compatible parameter type. In the current shape, i find them hidden and hard to reuse (I might be wrong).

>   The sum of the climb for all the line geometries in the layer.

Author: This is a processing algorithm, and all these output values can be accessed and used by programs that use the algorithm and they can also be utilised as inputs to other algorithms in the graphical modeller. All these outputs are declared in a standard way in the algorithm code.  This is how Processing algorithms work.
Since they are over-all values, it is not natural to include them in the output layer, and I don't think they should be provided as a separate "collection" output in addition to the individual outputs.

>   The sum of the climb for all the line geometries in the layer.

Reviewer: Thanks for the information. Never noticed these values in log panel beforehand. Do you know if there are other algs that report values in this way (i'm not sure we have documented any)
That said, should we not reuse the same label as in the output (eg, from memory, minelevation)?

>   The sum of the climb for all the line geometries in the layer.

Author: All processing algorithms report output in a result dictionary.  For many algorithms the result has only one entry (often called "OUTPUT").  An example of an algorithm that has several entries in the result is "Basic statistics for fields" - it has 18 values, I think.
Use the same names as in the result dictionary is probably a good idea.

>   The sum of the climb for all the line geometries in the layer.

Reviewer: > An example of an algorithm that has several entries in the result is "Basic statistics for fields"

result, which is formatted within a html file based (that I struggle to open - bug or me?) as I was suggesting earlier. Anyway, this is not the place to discuss implementation.
Do as you feel is better and coherent.

>   The sum of the climb for all the line geometries in the layer.

Author: Yes, the results are also returned in as HTML.  The point with returning the results in a dictionary is that they will be easier to get to by software that use the algorithm, and possible to use in the graphical modeller.

>   The sum of the climb for all the line geometries in the layer.

Author: I have added the result dictionary keys to the description of the outputs of the algorithm.

> ``Total descent``

Author: ```suggestion
``Total descent`` [number]
```

  The sum of the descent for all the line geometries in the layer.

> ``Minimum elevation``

Author: ```suggestion
``Minimum elevation`` [number]
```

  The minimum elevation for the geometries in the layer.

> ``Maximum elevation``

Author: ```suggestion
``Maximum elevation`` [number]
```
